android/src/main/java/io/ably/flutter/plugin/StreamsChannel.java

```diff
@@ -0,0 +1,205 @@
+import io.flutter.plugin.common.EventChannel;
+import io.flutter.plugin.common.MethodCall;
+import io.flutter.plugin.common.MethodCodec;
+
+
+final class StreamsChannel {
+
+    public interface StreamHandlerFactory {
+        EventChannel.StreamHandler create(Object arguments);
+    }
+
+    private static final String TAG = "StreamsChannel#";
+
+    private final BinaryMessenger messenger;
+    private final String name;
+    private final MethodCodec codec;
+    private IncomingStreamRequestHandler incomingStreamRequestHandler;
+
+    StreamsChannel(BinaryMessenger messenger, String name, MethodCodec codec) {
+        if (BuildConfig.DEBUG) {
+            if (messenger == null) {
+                Log.e(TAG, "Parameter messenger must not be null.");
+            }
+            if (name == null) {
+                Log.e(TAG, "Parameter name must not be null.");
+            }
+            if (codec == null) {
+                Log.e(TAG, "Parameter codec must not be null.");
+            }
+        }
+        this.messenger = messenger;
+        this.name = name;
+        this.codec = codec;
+    }
+
+    @UiThread
+    void setStreamHandlerFactory(final StreamHandlerFactory factory) {
+        incomingStreamRequestHandler = new IncomingStreamRequestHandler(factory);
+        messenger.setMessageHandler(name, incomingStreamRequestHandler);
+    }
+
+    void reset(){
+        incomingStreamRequestHandler.clearAll();
+    }
+
+    private final class IncomingStreamRequestHandler implements BinaryMessageHandler {
+        private final StreamHandlerFactory factory;
+        private final ConcurrentHashMap<Integer, StreamsChannel.Stream> streams = new ConcurrentHashMap<>();
+        private final ConcurrentHashMap<Integer, Object> listenerArguments = new ConcurrentHashMap<>();
+
+        IncomingStreamRequestHandler(StreamHandlerFactory factory) {
+            this.factory = factory;
+        }
+
+        @Override
+        public void onMessage(ByteBuffer message, final BinaryReply reply) {
+            final MethodCall call = codec.decodeMethodCall(message);
+            final String[] methodParts = call.method.split("#");
+
+            if (methodParts.length != 2) {
+                reply.reply(null);
+                return;
+            }
+
+            final int id;
+            try {
+                id = Integer.parseInt(methodParts[1]);
+            } catch (NumberFormatException e) {
+                reply.reply(codec.encodeErrorEnvelope("error", e.getMessage(), null));
+                return;
+            }
+
+            final String method = methodParts[0];
+            switch (method) {
+                case "listen":
+                    onListen(id, call.arguments, reply);
+                    break;
+                case "cancel":
+                    onCancel(id, call.arguments, reply);
+                    break;
+                default:
+                    reply.reply(null);
```

Comment:
I recognise this is code imported from elsewhere but this smells fishy... should that not be an exception thrown?

Reply:
@QuintinWillison there is no way of having another method type other than `"listen"` and `"cancel"` here. This code is borrowed from StreamsChannel with minor changes. Would you like to have this completely refactored now? or we can have a issue about this to track and deal with this later?

Reply:
Will it fail downstream when somebody unexpectedly receives `null` then?

Definitely no need to refactor now but I just want to make sure that this behaviour is documented and that somebody maintaining this months or years down the line will know that this is a downstream outcome.

Is this documented on the interface for `reply`'s `reply` method?

Reply:
I'll try by removing this and see while making other changes.

Reply:
@QuintinWillison after further investigation into this, I've observed that StreamsChannel has re-used the code that was implemented in flutter's default EventChannel. And the `reply.reply()` in the default case is also copied from the same. So, I'd say let's leave it that way. WDYT?

Reply:
And can you point out the inconsistency on usage of final?

Reply:
ok, as [I said before](https://github.com/ably/ably-flutter/pull/18#discussion_r450702980) no need to refactor... however I'm not sure what you're referring to when [you ask](https://github.com/ably/ably-flutter/pull/18#discussion_r453219515):

> And can you point out the inconsistency on usage of final?

Reply:
^ @tiholic I still don't understand to what you were referring to in [your last question](https://github.com/ably/ably-flutter/pull/18#discussion_r453219515).

Reply:
I mean if you can point out the inconsistency by sharing the line numbers, to make it a bit clearer for me. Anyhow as you said, we can ignore this.

Reply:
Oh, I get you now - you're referring the [my outer comment](https://github.com/ably/ably-flutter/pull/18#pullrequestreview-440140108).

I think the specific example I found was the [onMessage implementation](https://github.com/ably/ably-flutter/blob/d99a8e40f166333522d84a33c7ee7c23c98714bc/android/src/main/java/io/ably/flutter/plugin/StreamsChannel.java#L76), where for some non-apparent reason only the `reply` argument has been made `final`.

Nowhere else in that source file have arguments been made `final` which is inconsistent in and of itself as a single unit of source code.

I think we can ignore this for now as there's no real benefit on improving this right now. It may be something we can look for later on using lint / checkstyle. I'll resolve this thread. 😅